### PR TITLE
[flang][cuda] Add shape to cuf.data_transfer operation

### PR DESCRIPTION
**Pull request by clementval**
When doing data transfer with dynamic sized array, we are currently generating a data transfer between two descriptors. If the shape values can be provided, we can keep the data transfer between two references. This patch adds the shape operands to the operation.

This will be exploited in lowering in a follow up patch.

### CODE REVIEW

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -4444,7 +4445,9 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         !userDefinedAssignment) {
       Fortran::lower::StatementContext localStmtCtx;
       hlfir::Entity rhs = evaluateRhs(localStmtCtx);
+      llvm::errs() << rhs << "\n";
```
Comment (vzakhari) [inline flang/lib/Lower/Bridge.cpp:4448]: Leftovers.

File: flang/lib/Optimizer/Dialect/CUF/CUFOps.cpp [hunk 2]
```diff
@@ -99,6 +99,11 @@ llvm::LogicalResult cuf::AllocateOp::verify() {
 llvm::LogicalResult cuf::DataTransferOp::verify() {
   mlir::Type srcTy = getSrc().getType();
   mlir::Type dstTy = getDst().getType();
+  if (!getShape().empty()) {
+    if (!fir::isa_ref_type(srcTy) || fir::isa_ref_type(dstTy))
```
Comment (vzakhari) [inline flang/lib/Optimizer/Dialect/CUF/CUFOps.cpp:103]: ?

```suggestion
    if (!fir::isa_ref_type(srcTy) || !fir::isa_ref_type(dstTy))
```

File: flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td [hunk 3]
```diff
@@ -161,10 +161,11 @@ def cuf_DataTransferOp : cuf_Op<"data_transfer", []> {
 
   let arguments = (ins Arg<AnyType, "", [MemRead]>:$src,
                        Arg<AnyRefOrBoxType, "", [MemWrite]>:$dst,
+                       Variadic<AnyIntegerType>:$shape,
```
Comment (vzakhari) [inline flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td:164]: Is there a reason for not using a `!fir.shape` value here?

File: flang/lib/Lower/Bridge.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Lower/Bridge.cpp:4448]: Yes. Sorry about that.

File: flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td (hunk 3 above)
Comment (clementval) [reply flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td]: No special reason. I took the definition for the AllocOp but shape is fine. I updated the op.

**Review by vzakhari — APPROVED** [review]

**Pull request merged**